query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs

```diff
@@ -141,28 +141,21 @@ impl DataInputFieldMapper for UpdateDataInputFieldMapper {
         input_field(rf.name.clone(), InputType::object(input_object), None).optional()
     }
 
-    fn map_composite(&self, ctx: &mut BuilderContext, _cf: &CompositeFieldRef) -> InputField {
-        // Todo: Build composite types
-        // - Include scalar update operators (`set` value, `increment`, `decrement`, ...) (name clash potential? double check)
-        // - `set`, `push`
-        // - If `set` is used for a composite: We can simplify the input parsing by _NOT_ having any other nested operations below a set.
+    fn map_composite(&self, ctx: &mut BuilderContext, cf: &CompositeFieldRef) -> InputField {
+        // Shorthand object (equivalent to the "set" operation).
+        let shorthand_type = InputType::Object(composite_set_update_object_type(ctx, cf));
+        // Operation envelope object.
+        let envelope_type = InputType::Object(composite_update_envelope_object_type(ctx, cf));
 
-        // Dummy code to let the engine boot
-        let ident = Identifier::new("dummy".to_string(), PRISMA_NAMESPACE);
+        let mut input_types = vec![envelope_type, shorthand_type.clone()];
 
-        let obj = match ctx.get_input_type(&ident) {
-            Some(o) => o,
-            None => {
-                let input_object = Arc::new(input_object_type(
-                    ident.clone(),
-                    vec![input_field("foo", InputType::int(), None)],
-                ));
-                ctx.cache_input_type(ident, input_object.clone());
-                Arc::downgrade(&input_object)
-            }
-        };
+        if cf.is_list() {
+            input_types.push(InputType::list(shorthand_type));
+        }
 
-        input_field("foo", InputType::object(obj), None)
+        input_field(cf.name.clone(), input_types, None)
+            .nullable_if(!cf.is_required())
```

Comment:
I'm not 100% sure right now, but I believe this must also check for `!cf.is_list()`. I'm not sure how lists and field requiredness work together off the top of my head.

Reply:
Yeah, you're right. I added an `is_optional` helper that checks for `FieldArity::Optional`.